Model Bal Int as 64 bits in codegen phase

shubhamnarlawar opened the pull request:
Resolves #156

shubhamnarlawar commented:
I have changed from `LLVMInt32Type()` to `LLVMInt64Type()` in `Package.cpp` and `ConstLoadInsn.cpp`.

@ruvi-d - Please suggest on how to handle maps in this case.

ruvi-d commented:
> @ruvi-d - Please suggest on how to handle maps in this case.

Same as for array of int types. The corresponding runtime functions need to be updated.

shubhamnarlawar commented:
@ruvi-d - Please review and suggest changes.

ruvi-d commented:
Please:
* Add a test for int value that is larger than 32 bits
* Remove 32 bit print int functions from runtime lib (update the tests accordingly)

shubhamnarlawar commented:
> Please:
> 
>     * Add a test for int value that is larger than 32 bits
> 
>     * Remove 32 bit print int functions from runtime lib (update the tests accordingly)



> Please:
> 
>     * Add a test for int value that is larger than 32 bits
> 
>     * Remove 32 bit print int functions from runtime lib (update the tests accordingly)

1. I have added a test which has a int value `> 32 bits` - `greater_than_32_bits.bal`
2. I have replaced all `printu32()` with `printu64()` in all bal tests.
3. I am using `uint64_t` instead of `int` as `int` just hold 4 bytes. Is this fine?

ruvi-d commented:
> 1. I have added a test which has a int value `> 32 bits` - `greater_than_32_bits.bal`

In the same test, print both positive and negative values that are larger than 32bit

> 2. I have replaced all `printu32()` with `printu64()` in all bal tests.

We should be using only signed 64 values. i.e `print_i64` not `print_u64`. Ballerina int's are signed and 64bit

> 3. I am using `uint64_t` instead of `int` as `int` just hold 4 bytes. Is this fine?

Again, you can't use unsigned. I assume that `long long` or `int64_t` will work

ruvi-d commented on test/any/cast_any_basic_types.bal:
```diff
@@ -2,7 +2,7 @@
 public function print_string(string val) = external;
 
-public function printu32(int val) = external;
+public function printu64(int val) = external;
```
Function not in use. Can be deleted.

shubhamnarlawar replied:
Deleted all unsigned print int functions from rust.

ruvi-d commented on test/int/greater_than_32_bits.bal:
```diff
@@ -0,0 +1,12 @@
+public function printu64(int val) = external;
+
+public function main() {
+    int val = 8589934591;
```
Need to test a negative value too

shubhamnarlawar replied:
Done. I have added negative value also in `greater_than_32_bits.bal`

ruvi-d reviewed (changes requested):
* Need to change all u64 references to i64
* Need to add test cases to cover negative value ints
* Delete Rust print functions that we are not using

shubhamnarlawar commented:
> > 1. I have added a test which has a int value `> 32 bits` - `greater_than_32_bits.bal`
> 
> In the same test, print both positive and negative values that are larger than 32bit

Done.

> 
> > 1. I have replaced all `printu32()` with `printu64()` in all bal tests.
> 
> We should be using only signed 64 values. i.e `print_i64` not `print_u64`. Ballerina int's are signed and 64bit

Got it. We have `print64`. Let me rename it to `print_i64` and update all the testcases.

> 
> > 1. I am using `uint64_t` instead of `int` as `int` just hold 4 bytes. Is this fine?
> 
> Again, you can't use unsigned. I assume that `long long` or `int64_t` will work

`int64_t` seems fine to me.

shubhamnarlawar commented:
>     * Need to change all u64 references to i64

Done.

> 
>     * Need to add test cases to cover negative value ints

Done in `greater_than_32_bits.bal` and `init_int.bal`.

> 
>     * Delete Rust print functions that we are not using

Removed unsued rust functions.

shubhamnarlawar commented:
Is it good to merge now?

ruvi-d commented on compiler/BIRReader.h:
```diff
@@ -313,7 +313,7 @@ class ConstantPoolSet {
     ConstantPoolEntry *getEntry(int index) { return poolEntries[index].get(); }
     std::string getStringCp(uint32_t index);
-    uint32_t getIntCp(uint32_t index);
+    uint64_t getIntCp(uint32_t index);
```
Still using unsigned int

shubhamnarlawar replied:
During initial stage, @ayonam suggested to read all the values from the BIR dump as unsigned values and not signed values.

shubhamnarlawar replied:
All byte stream functions return unsigned values. As per your suggestion to store it as `int64_t`, I am doing casting from `uint64_t` to `int64_t`.

ruvi-d replied:
BIR spec says that it's 8 bytes and signed : https://github.com/ballerina-platform/ballerina-lang/blob/nballerina/docs/bir-spec/src/main/resources/kaitai/bir.ksy#L49

shubhamnarlawar replied:
Done. I have updated all the places. Thanks for the suggestion.

ruvi-d commented on runtime/src/lib.rs:
```diff
@@ -86,49 +86,25 @@ pub extern "C" fn is_same_type(src_type: *const c_char, dest_type: *const c_char
 // Prints 32 bit signed integer
 #[no_mangle]
-pub extern "C" fn print32(num32: i32) {
+pub extern "C" fn print_i32(num32: i32) {
```
There is no need to use this function anymore

shubhamnarlawar replied:
Removed.

ruvi-d commented on runtime/src/lib.rs:
```diff
@@ -86,49 +86,25 @@ pub extern "C" fn is_same_type(src_type: *const c_char, dest_type: *const c_char
 // Prints 16 bit signed integer
 #[no_mangle]
-pub extern "C" fn print16(num16: i16) {
+pub extern "C" fn print_i16(num16: i16) {
```
We don't need this either right?

shubhamnarlawar replied:
No. All functions can use `print_i64`.
Removed.

ruvi-d commented on runtime/src/lib.rs:
```diff
@@ -86,49 +86,25 @@ pub extern "C" fn is_same_type(src_type: *const c_char, dest_type: *const c_char
-// Prints 8 bit unsigned integer
-#[no_mangle]
-pub extern "C" fn printu8(num8: u8) {
+pub extern "C" fn print_i8(num8: i8) {
```
Do we need print_i8?

shubhamnarlawar replied:
Removed.

ruvi-d commented on runtime/src/lib.rs:
```diff
@@ -86,49 +86,25 @@ pub extern "C" fn is_same_type(src_type: *const c_char, dest_type: *const c_char
 // Prints 64 bit signed integer
 #[no_mangle]
-pub extern "C" fn print64(num64: i64) {
+pub extern "C" fn print_i64(num64: i64) {
```
`print_integer` would have been a more appropriate name here

shubhamnarlawar replied:
As `print_i32`, `print_i16` and `print_i8` are removed. `print_integer` seems a good name. Thanks for the suggestion.

ruvi-d reviewed (changes requested):
One issue to resolve: https://github.com/ballerina-platform/nballerina/pull/235#discussion_r607951884

shubhamnarlawar commented:
> One issue to resolve: [#235 (comment)](https://github.com/ballerina-platform/nballerina/pull/235#discussion_r607951884)

Done.

shubhamnarlawar commented:
Please approve this PR - @Kishanthan @uthaiyashankar @manuranga @KavinduZoysa

shubhamnarlawar commented:
Merging PR #196 causes merge conflict in this PR because both PR were related to changes in all the testcases.
